MapboxGLAndroidSDK/src/cpp/style/sources/source.cpp

```diff
@@ -50,205 +55,263 @@ static std::unique_ptr<Source> createSourcePeer(jni::JNIEnv& env,
     }
 }
 
-    const jni::Object<Source>& Source::peerForCoreSource(jni::JNIEnv& env, mbgl::style::Source& coreSource, AndroidRendererFrontend& frontend) {
-        if (!coreSource.peer.has_value()) {
-            coreSource.peer = createSourcePeer(env, coreSource, &frontend);
-        }
-        return coreSource.peer.get<std::unique_ptr<Source>>()->javaPeer;
+const jni::Object<Source> &
+Source::peerForCoreSource(jni::JNIEnv &env, mbgl::style::Source &coreSource,
+                          AndroidRendererFrontend &frontend) {
+    if (!coreSource.peer.has_value()) {
+        coreSource.peer = createSourcePeer(env, coreSource, &frontend).release();
     }
+    return coreSource.peer.get<Source*>()->javaPeer;
+}
 
-    const jni::Object<Source>& Source::peerForCoreSource(jni::JNIEnv& env, mbgl::style::Source& coreSource) {
-        if (!coreSource.peer.has_value()) {
-            coreSource.peer = createSourcePeer(env, coreSource, nullptr);
-        }
-        return coreSource.peer.get<std::unique_ptr<Source>>()->javaPeer;
+const jni::Object<Source> &
+Source::peerForCoreSource(jni::JNIEnv &env, mbgl::style::Source &coreSource) {
+    if (!coreSource.peer.has_value()) {
+        coreSource.peer = createSourcePeer(env, coreSource, nullptr).release();
     }
+    return coreSource.peer.get<Source*>()->javaPeer;
+}
 
-    Source::Source(jni::JNIEnv& env,
-                   mbgl::style::Source& coreSource,
-                   const jni::Object<Source>& obj,
-                   AndroidRendererFrontend* frontend)
-        : source(coreSource), javaPeer(jni::NewGlobal(env, obj)), rendererFrontend(frontend) {}
+Source::Source(jni::JNIEnv &env,
+               mbgl::style::Source &coreSource,
+               const jni::Object<Source> &obj,
+               AndroidRendererFrontend *frontend)
+        : source(coreSource.makeWeakPtr()), javaPeer(jni::NewGlobal(env, obj)), rendererFrontend(frontend) {}
 
-    Source::Source(jni::JNIEnv&, std::unique_ptr<mbgl::style::Source> coreSource)
-        : ownedSource(std::move(coreSource))
-        , source(*ownedSource) {
+Source::Source(jni::JNIEnv &, std::unique_ptr<mbgl::style::Source> coreSource)
+        : ownedSource(std::move(coreSource)), source(ownedSource->makeWeakPtr()) {
+}
+
+Source::~Source() {
+    if (ownedSource) {
+        ownedSource.reset();
+        ownedSource.release();
```

Comment:
not sure what is expected here? reset deletes owned object and sets current ptr to nullptr, release can be omitted.

Reply:
Right,